app/scripts/controllers/app-state-controller.test.ts

```diff
@@ -43,680 +46,648 @@ const extensionMock = {
+            if (typeof state.surveyLinkLastClickedOrClosed === 'number') {
+              controller.setSurveyLinkLastClickedOrClosed(
+                state.surveyLinkLastClickedOrClosed,
+              );
+            }
+          },
+        );
+
+        controllerMessenger.publish(
+          'AppStateController:stateChange',
+          {
+            surveyLinkLastClickedOrClosed: timeNow,
+          } as unknown as AppStateControllerState,
+          [],
+        );
+
+        expect(controller.state.surveyLinkLastClickedOrClosed).toStrictEqual(
+          timeNow,
+        );
+        expect(
+          controllerMessenger.call('AppStateController:getState')
+            .surveyLinkLastClickedOrClosed,
+        ).toStrictEqual(timeNow);
+      });
+    });
 
-      expect(
-        appStateController.store.getState().surveyLinkLastClickedOrClosed,
-      ).toStrictEqual(timeNow);
-      expect(
-        controllerMessenger.call('AppStateController:getState')
-          .surveyLinkLastClickedOrClosed,
-      ).toStrictEqual(timeNow);
+    it('state will be published when there is state change', async () => {
+      await withController(({ controller, controllerMessenger }) => {
+        expect(controller.state.surveyLinkLastClickedOrClosed).toStrictEqual(
+          null,
+        );
+        const timeNow = Date.now();
+        controllerMessenger.subscribe(
+          'AppStateController:stateChange',
+          (state: Partial<AppStateControllerState>) => {
+            expect(state.surveyLinkLastClickedOrClosed).toStrictEqual(timeNow);
+          },
+        );
+
+        controller.setSurveyLinkLastClickedOrClosed(timeNow);
+
+        expect(controller.state.surveyLinkLastClickedOrClosed).toStrictEqual(
+          timeNow,
+        );
+        expect(
+          controllerMessenger.call('AppStateController:getState')
+            .surveyLinkLastClickedOrClosed,
+        ).toStrictEqual(timeNow);
+      });
     });
   });
 });
+
+type WithControllerOptions = Partial<AppStateControllerOptions>;
+
+type WithControllerCallback<ReturnValue> = ({
+  controller,
+  controllerMessenger,
+}: {
+  controller: AppStateController;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  controllerMessenger: any;
+}) => ReturnValue;
+
+type WithControllerArgs<ReturnValue> =
+  | [WithControllerCallback<ReturnValue>]
+  | [WithControllerOptions, WithControllerCallback<ReturnValue>];
+
+async function withController<ReturnValue>(
+  ...args: WithControllerArgs<ReturnValue>
+): Promise<ReturnValue> {
+  const [options = {}, fn] = args.length === 2 ? args : [{}, args[0]];
+
+  const controllerMessenger = new ControllerMessenger<
+    | AppStateControllerActions
+    | AddApprovalRequest
+    | AcceptRequest
+    | PreferencesControllerGetStateAction,
+    | AppStateControllerEvents
+    | PreferencesControllerStateChangeEvent
+    | KeyringControllerQRKeyringStateChangeEvent
+  >();
+  jest.spyOn(ControllerMessenger.prototype, 'call');
```

Comment:
Is it necessary to spy on these methods? Usually we don't need to mock any of the messenger API.

Reply:
Yes I agree I did my best to align with how we test other controllers but this was there before

Reply:
Ah yes, sorry, no problem then.

Reply:
Spy removed